tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,271 @@
+      this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/device.fwd.h"
+
+namespace tiny_dnn {
+
+/* Class modeling a Tensor
+ * @param b The size of the batch
+ * @param w The width of the tensor
+ * @param h The height of the tensor
+ * @param d The number of channels
+ */
+template<class U = float_t,
+	 cnn_size_t b = 1,
+	 cnn_size_t w = 1,
+	 cnn_size_t h = 1,
+	 cnn_size_t d = 1>
+class Tensor {
+ public:
+    /* Default constructor for the Tensor class
+     */
+    Tensor() : shape_(4) {
+        init_data (b, w, h, d);
+        init_shape(b, w, h, d);
+    }
+
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+    typedef CLCudaAPI::Buffer<U>    buffer;
+    typedef std::unique_ptr<buffer> buffer_ptr;
+#endif
+
+    typedef std::unique_ptr<std::vector<U> > data_ptr;
+    
+    // Returns the tensor shape
+    std::vector<cnn_size_t> shape() const { return shape_; }
+    
+    // Returns the value of a specified index in the tensor.
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T& at(const cnn_size_t batch,
+          const cnn_size_t width,
+          const cnn_size_t height,
+          const cnn_size_t depth) {
+        return *access_data<T>(batch, width, height, depth);
+    }
+
+    template<typename T>
+    const T& at(const cnn_size_t batch,
+                const cnn_size_t width,
+                const cnn_size_t height,
+                const cnn_size_t depth) const {
+        return *access_data<T>(batch, width, height, depth);
+    }
+
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T* ptr(const cnn_size_t batch,
+           const cnn_size_t width,
+           const cnn_size_t height,
+           const cnn_size_t depth) {
+        return access_data<T>(batch, width, height, depth);
+    }
+    
+    template<typename T>
+    const T* ptr(const cnn_size_t batch,
+                 const cnn_size_t width,
+                 const cnn_size_t height,
```

Comment:
If you rename `width` to `x` and `height` to `y`, all the problems in the world will not be solved for sure.